src/cocoa/toga_cocoa/widgets/numberinput.py

```diff
@@ -68,48 +27,96 @@ def controlTextDidEndEditing_(self, notification) -> None:
-    def __init__(self, id=None, style=None, factory=None,
-                 min_value=0, max_value=100, step=1):
-        self.min_value = min_value
-        self.max_value = max_value
-        self.step = step
-        super().__init__(id=id, style=style, factory=factory)
-        self._impl = Stepper(interface=self)
+class NumberInput(Widget):
+    def create(self):
+        self.native = NSTextField.new()
+        self.native.interface = self.interface
+
+        delegate = TogaNumberFieldDelegate.new()
+        delegate.interface = self.interface
+        delegate.controller = self.native
+        self.native.delegate = delegate
+
+        self.native.bezeled = True
+        self.native.bezelStyle = NSTextFieldSquareBezel
 
-    @property
-    def value(self):
-        return self._impl.native.floatValue
+        # Add the layout constraints
+        self.add_constraints()
 
-    @value.setter
-    def value(self, val):
-        self._impl.native.floatValue = float(val)
+    def set_readonly(self, value):
+        # Even if it's not editable, it's still selectable.
+        self.native.editable = not value
+        self.native.selectable = True
 
-    def on_change(self, handler):
-        self.controller.stepper_update()
+    def set_placeholder(self, value):
+        self.native.cell.placeholderString = value
 
+    def set_alignment(self, value):
+        self.native.alignment = NSTextAlignment(value)
 
-class Stepper(Widget):
-    def create(self):
-        self.native = TogaStepper.alloc().init()
-        self.native.interface = self.interface
+    def set_font(self, value):
+        if value:
+            self.native.font = value._impl.native
 
-        self.native.minValue = self.interface.min_value
-        self.native.maxValue = self.interface.max_value
-        self.native.increment = self.interface.step
-        self.native.target = self.native
-        self.native.action = SEL('onChange:')
+    def get_value(self):
+        return self.native.stringValue
 
-        self.add_constraints()
+    def set_value(self, value):
+        self.native.stringValue = value
 
     def rehint(self):
-        fitting_size = self.native.fittingSize()
-        self.interface._intrinsic.width = fitting_size.width
-        self.interface._intrinsic.height = fitting_size.height
-        # self.interface.style.margin_top = -3
+        # Height of a text input is known and fixed.
+        # Width must be > 100
+        # print("REHINT TextInput", self, self._impl.fittingSize().width, self._impl.fittingSize().height)
+        self.interface.intrinsic.width = at_least(self.interface.MIN_WIDTH)
+        self.interface.intrinsic.height = self.native.fittingSize().height
 
+    def set_on_change(self, handler):
+        pass
 
-class TogaStepper(NSStepper):
-    @objc_method
-    def onChange_(self, obj) -> None:
-        process_callback(self.interface.on_change(self.interface))
+
+# class StepperInterface(InterfaceWidget):
```

Comment:
Feels a little strange to have this big block of commented code here. We'd probably reject if this were a non-core contributor.

Reply:
That's a fair cop :-) I'll get this cleaned up before I merge. Strictly, it's a small functionality regression anyway (although the widget hasn't worked *at all* since the impl/native restructure).